unit_tests/test_redis_client_base.py

```diff
@@ -0,0 +1,235 @@
+
+    @staticmethod
+    @patch("uuid.uuid4", return_value="some-transaction-uuid")
+    def test_unregister(mock_uuid, redis_client_auto_register):
+        """Check whether the redis.db.publish is called with correct arguments."""
+        redis_client_auto_register.is_active = True
+        redis_client_auto_register.unregister(is_blocking=False)
+        data = {"name": AREA_ID, "transaction_id": mock_uuid()}
+        redis_client_auto_register.redis_db.publish.assert_called_once_with(
+            f"{AREA_ID}/unregister_participant",
+            json.dumps(data))
+
+    @staticmethod
+    def test_unregister_is_active_false_throws_exception(redis_client_auto_register):
+        """Check if is active set to false throws exception when user is already unregistered."""
+        with pytest.raises(RedisAPIException,
+                           match="API is already unregistered from the market."):
+            redis_client_auto_register.is_active = False
+            redis_client_auto_register.unregister(is_blocking=True)
+
+    @staticmethod
+    @patch("d3a_api_client.redis_client_base.wait_until_timeout_blocking",
+           side_effect=AssertionError)
+    def test_unregister_is_active_true_throws_exception(
+            mock_wait_until_timeout_blocking,
+            redis_client_auto_register):
+        """Check if is active set to true throws exception."""
+        with pytest.raises(RedisAPIException,
+                           match="API unregister process timed out. "
+                                 "Server will continue processing "
+                                 "your request on the background and "
+                                 "will notify you as soon as "
+                                 "the unregistration has been completed."):
+            redis_client_auto_register.is_active = True
+            redis_client_auto_register.unregister(is_blocking=True)
+        mock_wait_until_timeout_blocking.assert_called()
+
+    @staticmethod
+    @patch("uuid.uuid4", return_value="some-transaction-uuid")
+    def test_on_register(mock_uuid, redis_client_auto_register):
+        """Check the on_register function with correct message that doesn't throw exception."""
+        redis_client_auto_register.register(is_blocking=False)
+        data = {"name": AREA_ID, "device_uuid": DEVICE_ID, "transaction_id": mock_uuid()}
+        message = {"data": json.dumps(data)}
+        redis_client_auto_register._on_register(message)
+        assert redis_client_auto_register.is_active is True
+        assert redis_client_auto_register.area_uuid == DEVICE_ID
+
+    @staticmethod
+    @patch("uuid.uuid4", return_value="some-transaction-uuid")
+    def test_on_unregister(mock_uuid, redis_client_auto_register):
+        """Check the on_unregister function with correct message that doesn't throws exception."""
+        redis_client_auto_register.is_active = True
+        redis_client_auto_register.unregister(is_blocking=False)
```

Comment:
I would advise not calling the unregister method here, you can mock the `_check_buffer_message_matching_command_and_id` and directly call the _on_unregister  method (Unless we specifically agreed to not test private methods)

Reply:
As you said, the reason why we decided not to call the `_check_buffer_message_matching_command_and_id` method is that it is private. Ideally, we shouldn't know anything about it (from the perspective of the test). What we know, instead, is the public interface of the class and that's why we opted to use it here to "set up" the conditions for the test.